state/txindex/kv/kv.go

```diff
@@ -40,6 +47,101 @@ type TxIndex struct {
 	log log.Logger
 }
 
+func (txi *TxIndex) Prune(retainHeight int64) (int64, int64, error) {
+	// Returns the last retained height
+	lastRetainHeight, err := txi.getIndexerRetainHeight()
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to look up last block indexer retain height: %w", err)
+	}
+	if lastRetainHeight == 0 {
+		lastRetainHeight = 1
+	}
+
+	ctx := context.Background()
+	results, err := txi.Search(ctx, query.MustCompile(
+		fmt.Sprintf("tx.height < %d AND tx.height >= %d", retainHeight, lastRetainHeight)))
+	if err != nil {
+		panic(err)
+	}
+	if len(results) == 0 {
+		return 0, lastRetainHeight, nil
+	}
+
+	batch := txi.store.NewBatch()
+	defer func(batch dbm.Batch) {
+		err := batch.Close()
+		if err != nil {
+			txi.log.Error(fmt.Sprintf("Error when closing pruning batch: %v", err))
+		}
+	}(batch)
+
+	sort.Sort(TxResultByHeight(results))
+	for _, result := range results {
+		errDeleteResult := txi.deleteResult(result, batch)
+		if errDeleteResult != nil {
+			// If we crashed in the middle of pruning the height,
+			// we assume this height is retained
+			errSetLastRetainHeight := txi.setIndexerRetainHeight(result.Height, batch)
+			if errSetLastRetainHeight != nil {
+				return 0, lastRetainHeight, fmt.Errorf("error setting last retain height '%v' while handling result deletion error '%v'", errSetLastRetainHeight, errDeleteResult)
+			}
+			errWriteBatch := batch.WriteSync()
+			if errWriteBatch != nil {
+				return 0, lastRetainHeight, fmt.Errorf("error writing batch '%v' while handling result deletion error '%v'", errWriteBatch, errDeleteResult)
+			}
+			return result.Height - lastRetainHeight, result.Height, errDeleteResult
+		}
```

Comment:
Maybe add a section here to flush the batch every X results. See https://github.com/cometbft/cometbft/blob/6a7dfb40d0302de5cdbca51cc4daede74b6d44c2/store/store.go#L312 which is called by https://github.com/cometbft/cometbft/blob/6a7dfb40d0302de5cdbca51cc4daede74b6d44c2/store/store.go#L379